cmd/data-usage.go

```diff
@@ -0,0 +1,205 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"math/rand"
+	"time"
+)
+
+const (
+	dataUsageSweepInterval   = 12 * time.Hour
+	dataUsageRefreshInterval = 10 * time.Minute
+)
+
+// A safe copy of a data usage info struct
+func copyDataUsageInfo(dataUsage DataUsageInfo) DataUsageInfo {
+	dataUsageCopy := DataUsageInfo{}
+	dataUsageCopy.ObjectsCount = dataUsage.ObjectsCount
+	dataUsageCopy.ObjectsTotalSize = dataUsage.ObjectsTotalSize
+	dataUsageCopy.BucketsCount = dataUsage.BucketsCount
+	dataUsageCopy.LastUpdate = dataUsage.LastUpdate
+	dataUsageCopy.ObjectsSizesHistogram = make(map[string]uint64)
+	for k, v := range dataUsage.ObjectsSizesHistogram {
+		dataUsageCopy.ObjectsSizesHistogram[k] = v
+	}
+	dataUsageCopy.BucketsSizes = make(map[string]uint64)
+	for k, v := range dataUsage.BucketsSizes {
+		dataUsageCopy.BucketsSizes[k] = v
+	}
+
+	return dataUsageCopy
+}
+
+// Safe copy of the global data usage info
+func safeCopyDataUsageInfo() DataUsageInfo {
+	globalDataUsageInfoMu.RLock()
+	defer globalDataUsageInfoMu.RUnlock()
+
+	return copyDataUsageInfo(globalDataUsageInfo)
+}
+
+func initDataUsageStats() {
+	go runDataUsageInfoUpdateRoutine()
+}
+
+func runDataUsageInfoUpdateRoutine() {
+	// Wait until the object layer is ready
+	var objAPI ObjectLayer
+	for {
+		objAPI = newObjectLayerWithoutSafeModeFn()
+		if objAPI == nil {
+			time.Sleep(time.Second)
+			continue
+		}
+		break
+	}
+
+	ctx := context.Background()
+
+	zones, ok := objAPI.(*xlZones)
+	if !ok {
+		runDataUsageInfoForFS(ctx, objAPI)
+	} else {
+		runDataUsageInfoForXL(ctx, zones)
+	}
+}
+
+func runDataUsageInfoForFS(ctx context.Context, objAPI ObjectLayer) {
+	for {
+		usageInfo := objAPI.DataUsageInfo(ctx)
+
+		globalDataUsageInfoMu.Lock()
+		globalDataUsageInfo = usageInfo
+		globalDataUsageInfoMu.Unlock()
+
+		time.Sleep(dataUsageSweepInterval)
+	}
+}
+
+func runDataUsageInfoForXL(ctx context.Context, zones *xlZones) {
+	for {
+		locker := zones.NewNSLock(ctx, minioMetaBucket, "leader-data-usage-info")
+		err := locker.GetLock(newDynamicTimeout(time.Millisecond, time.Millisecond))
```

Comment:
where is the unlock?

Reply:
Well, the unlock is not needed since the node, if elected as leader, will be always leader until killed or restarted.

Reply:
There should be an unlock otherwise we are purposefully creating stale locks.  Unlock should happen upon shutdown signal perhaps for all leader elections.

Also I think we should perhaps only do single leader election for all features including lifecycle, heal and accounting. This way single scan is necessary for each I/O. Also new leader should be selected once a cycle is over.

This will avoid IMO unnecessary lock requests allowing for single leader master at any given point in time.

Reply:
> This will avoid IMO unnecessary lock requests allowing for single leader master at any given point in time.

Well, we probably should distribute tasks to multiple nodes randomly so one node won't do everything

Reply:
Unlock still needs to happen, it's not updated in the PR.

Reply:
> Unlock still needs to happen, it's not updated in the PR.

Unlocking is done in L102 now